### PR TITLE
Fully Qualified Type names for the Source Generator (to avoid namespace clashes)

### PR DESCRIPTION
**Pull request by thomhurst**
I've got a solution where I'm getting compilation errors because of namespace naming conflicts. Fully qualifying the types in generated code (so they begin with `global::`) should fix this.

### CODE REVIEW

File: OneOf.SourceGenerator/OneOfGenerator.cs [hunk 1]
```diff
@@ -89,8 +89,8 @@ partial class {classNameWithGenericTypes}");
             foreach (var (param, arg) in paramArgPairs)
```
Comment (romfir) [inline OneOf.SourceGenerator/OneOfGenerator.cs:89]: shouldn't we also update similar usage in line 80?

**Comment by romfir** [comment]
@thomhurst  Please run OneOf.SourceGenerator.AnalyzerTests and fix them (I think they should now fail), could you also add "global::" prefix to `AttributeNamespace` and remove "global::" from filtering inside `GetSemanticTargetForGeneration`?
@mcintyre321 could we possibly run these tests in CI?

File: OneOf.SourceGenerator/OneOfGenerator.cs (hunk 1 above)
Comment (thomhurst) [reply OneOf.SourceGenerator/OneOfGenerator.cs:89]: Nope namespaces don't support `global::` - Only types do

**Comment by thomhurst** [comment]
> @thomhurst Please run OneOf.SourceGenerator.AnalyzerTests and fix them (I think they should now fail) and add a test for your specific case (for the error that you encountered), could you also add "global::" prefix to `AttributeNamespace` and remove "global::" from filtering inside `GetSemanticTargetForGeneration`? @mcintyre321 could we possibly run these tests in CI?

The comparison inside `GetSemanticTargetForGeneration` is just with strings, so I have to prepend `global::` there. But we're comparing it anyway against a fully qualified type, so we know we're getting the right one.

**Comment by thomhurst** [comment]
Actually I've just added some types with conflicting namespaces. If you put those types in a project with the existing code, it won't compile. So the fact that it compiles is proof it fixes it.

**Comment by thomhurst** [comment]
@romfir They have also run on CI here: 
https://ci.appveyor.com/project/mcintyre321/oneof/builds/47938445

File: OneOf.SourceGenerator/OneOfGenerator.cs (hunk 1 above)
Comment (romfir) [reply OneOf.SourceGenerator/OneOfGenerator.cs:89]: right, sorry I thought it can also be used for namespaces

**Comment by romfir** [comment]
Interesting I thought that [this](https://github.com/mcintyre321/OneOf/blob/bfadb1159677a06429a7a9eb54bac2b22d14eb24/OneOf.SourceGenerator.AnalyzerTests/AnalyzerTests.cs#L54) test would fails because of the missing namespace in implicit operators

**Comment by romfir** [comment]
`GeneratesPublicClass` fails on my machine with `(4,11): error CS7000: Unexpected use of an aliased name`

**Comment by thomhurst** [comment]
> Interesting I thought that [this](https://github.com/mcintyre321/OneOf/blob/bfadb1159677a06429a7a9eb54bac2b22d14eb24/OneOf.SourceGenerator.AnalyzerTests/AnalyzerTests.cs#L54) test would fails because of the missing namespace in implicit operators

Looks like the assertions aren't doing full equivalency. It seems to be doing things like checking no diagnostics, or that certain elements match?

**Comment by thomhurst** [comment]
> `GeneratesPublicClass` fails on my machine with `(4,11): error CS7000: Unexpected use of an aliased name`

You haven't changed the namespace to fully qualified have you?

**Comment by romfir** [comment]
> > Interesting I thought that [this](https://github.com/mcintyre321/OneOf/blob/bfadb1159677a06429a7a9eb54bac2b22d14eb24/OneOf.SourceGenerator.AnalyzerTests/AnalyzerTests.cs#L54) test would fails because of the missing namespace in implicit operators
> 
> Looks like the assertions aren't doing full equivalency. It seems to be doing things like checking no diagnostics, or that certain elements match?

They are using `SyntaxTree.IsEquivalentTo`, perhabs adding "global::" is a `trivia difference` mentioned in the method description 😄 also the comments seems to be ignored

>You haven't changed the namespace to fully qualified have you?

I did 🥲 

Could you move this new test to `AnalyzerTests`? All `OneOf.SourceGenerator.Tests` should be moved there but I did not have time to do it, perhaps the name is a bit misleading but its a better way to test source generators.

**Comment by thomhurst** [comment]
Hmm it's not a test as such, but just creating a type that'd fail to compile before. But to do that, I need access to the `GenerateOneOfAttribute` which the Analyzer project doesn't have

**Comment by thomhurst** [comment]
@romfir Have added a test to the analyzer project `GeneratesClassWithConflictingNamespaces`

**Comment by thomhurst** [comment]
@mcintyre321 Any change of getting this in please? I'm blocked from using the source generator at the moment because of this

**Review by romfir — APPROVED** [review]
LGTM!

**Pull request merged**